contracts/epochs/EpochClusterQCs.cdc

```diff
@@ -0,0 +1,201 @@
+
+// Manages the process of collecting votes for the root quorum certificate of the upcoming
+// epoch for all collection node clusters assigned for the upcoming epoch.
+//
+// This contract is a member of a series of epoch smart contracts which coordinates the 
+// process of transitioning between epochs in Flow.
+pub contract EpochClusterQCs {
+
+    // ================================================================================
+    // CONTRACT VARIABLES
+    // ================================================================================
+
+    // Indicates whether votes are currently being collected.
+    // If false, no node operator will be able to submit votes
+    pub var inProgress: Bool
+
+    // The collection node clusters for the current epoch
+    access(account) var clusters: [Cluster]
+
+    // Votes that nodes claim at the beginning of each EpochSetup phase
+    access(account) var generatedVotes: {String: Vote}
+
+    // Votes submitted per cluster
+    access(account) var votesByCluster: {UInt16: [Vote]}
+
+    // Indicates if a voter resource has already been claimed by a node ID
+    access(account) var voterClaimed: {String: Bool}
+
+    // ================================================================================
+    // CONTRACT CONSTANTS
+    // ================================================================================
+
+    // Canonical paths for admin and voter resources
+    pub let AdminStoragePath: Path
+    pub let VoterStoragePath: Path
+
+    // Represents a collection node cluster for a given epoch. 
+    pub struct Cluster {
+
+        // The IDs of the nodes in the cluster.
+        pub let nodeIDs: [String]
+
+        // The index of the cluster within the cluster assignment. This uniquely identifies
+        // a cluster for a given epoch
+        pub let index: UInt16
+
+        pub fun size(): UInt16 {
+            return UInt16(self.nodeIDs.length) 
+        }
+
+        // Returns the minimum number of votes required in order to be able to generate a
+        // valid quorum certificate for this cluster.
+        pub fun voteThreshold(): UInt16 {
+            return self.size() * UInt16(2) / UInt16(3) + UInt16(1)
+        }
+
+        init(index: UInt16, nodeIDs: [String]) {
+            self.index = index
+            self.nodeIDs = nodeIDs
+        }
+    }
+
+    // Vote represents a vote from one collection node. It simply contains a string with an
+    // encoded representation of the vote. Votes are aggregated to build quorum certificates;
+    // eventually we may want to do the aggregation and validate votes within the smart
+    // contract, but in the meantime the vote contents are opaque here.
+    pub struct Vote {
+        pub var nodeID: String
+        pub(set) var raw: String?
+        pub(set) var clusterIndex: UInt16
```

Comment:
I don't think this should be settable after instantiation. The `Voter` shouldn't be able to submit votes for a cluster they aren't a part of.

Reply:
yes, you're right. It should be a constant